### DISCUSSION TITLE
get UIC for an option

### DISCUSSION THREAD

**Discussion by richy1996** [hootnot/saxo_openapi#29] [General]
how do i provide the information to `InstrumentToUic` function to get the uic for an option, specifically the instrument information within the spec dictionary?

or if there's a way to specify a strike, put/call and expiration to `ContractoptionSpaces` function that would work too i guess.

As an example, I'd like to get the uic for the SPXW (note there's a difference between SPXW and SPX) option, asset type is "StockIndexOption", strike of 5130, Put, expiration date of March 5, 2024. The underlying uic is 1909049.

Any help is appreciated.

**Comment by richy1996** [comment]
bit more info, the option root id should be 128.

**Comment by richy1996** [comment]
current workaround seems ok i guess
```python
import saxo_openapi.endpoints.referencedata as rd

def return_uic_option(OptionRootId, strike, put_call, expiration):
    params = {
        "ExpiryDates": expiration,
        "OptionSpaceSegment": "SpecificDates",
    }
    r = rd.instruments.ContractoptionSpaces(
                             OptionRootId=OptionRootId,
                             params=params)
    response = client.request(r)
    
    for option_space in response.get('OptionSpace'):
        if option_space['Expiry'] == expiration:
            for specific_option in option_space.get('SpecificOptions'):
                if specific_option['PutCall'] == put_call and specific_option['StrikePrice'] == strike:
                    return specific_option

OptionRootId = 128
strike = 5130
put_call = 'Put'
expiration = '2024-03-05'

return_uic_option(OptionRootId, strike, put_call, expiration)
# {'PutCall': 'Put',
#  'StrikePrice': 5130.0,
#  'TradingStatus': 'Tradable',
#  'Uic': 40381718,
#  'UnderlyingUic': 1909049}
```

**Reply by hootnot** [reply]
Hi, pls. check the code in the Jupyter notebooks. It contains option / referencedata related example code.

**Reply by richy1996** [reply]
the [instrumentlookup.ipynb](https://github.com/hootnot/saxo_openapi/blob/503735ae06eeb05f7740ea31f3b45be3e67dd538/jupyter/instrumentlookup.ipynb) has a bit of info on options but no more than what I've shared above. maybe I'm missing one but as far as I can tell, none simply show you how to get a uic from attributes of an option like underlying, strike, expiration, call/put.

**Comment by armanobosyan** [comment]


**Reply by richy1996** [reply]
can you show an example using an option? I only see an example for AAPL and it's not clear to me how to easily convert that to be usable for an option.

**Reply by richy1996** [reply]
are you serious?

**Reply by richy1996** [reply]
or maybe you're trying to tell me i've already found the best way to get the uic

**Reply by hootnot** [reply]
The [instrumentlookup.ipynb](https://github.com/hootnot/saxo_openapi/blob/503735ae06eeb05f7740ea31f3b45be3e67dd538/jupyter/instrumentlookup.ipynb) contains an example of Royal Dutch Shell. That should be the pattern 

You can simplify that by wrapping things in code the way you want it.
But it involves the 2 steps as you already did yourself. 

Figure out the`OptionRootId`

```
r = rd.instruments.Instruments(params = {'AccountKey': ai.AccountKey,
                                         'KeyWords': 'Shell',
                                         'AssetTypes': AT.StockOption})
print(json.dumps(client.request(r), indent=2))
```
and then query the Options:

```
OptionRootId = 231
r = rd.instruments.ContractoptionSpaces(OptionRootId=OptionRootId)
client.request(r)
print(json.dumps(r.response, indent=4))
```
Process the JSON the way you need it.

**Reply by richy1996** [reply]
ok got it, thx for confirming
